### PR TITLE
Populate yt.loaders module (followup #2831)

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary

Follow up to #2831: move loader functions to yt.loaders and redirect internal imports
(note that this could be merged first now that it is stabilised, making #2831 redundant)

As a side benefit, I slightly refactored the load_sample function (gave more sensible names to some inner variables and a helper class).
Another side effect is the removal of the `refine_amr` function, which implementation was very tricky to isolate without creating circular imports, and which apparently fell out of use. This removal and that of the affected test lines is the reason for the significant line count diff (~ -220 lines).

## PR Checklist

- [x] pass `black --check yt/`
- [x] pass `isort . --check --diff`
- [x] pass `flake8 yt/`
- [x] isolate stream.data_stuctures.py content in 3 files
- [x] relocate stream loaders to yt.loaders (refactoring needed to avoid circular imports)
- [x] relocate convenience loaders (trivial)
- [x] relocate load_sample
- [x] ~relocate all load tests to yt/tests/test_loaders.py~ I now think it's okay to leave them where they are actually
- [x] cleanup stream.api
- [x] update documentation references
- [x] reactivate flake8 for yt.loaders

### CODE REVIEW

**Comment by matthewturk** [comment]
I'm okay with removing `refine_amr` but I want to get @jzuhone to sign off, too.

**Comment by neutrinoceros** [comment]
> I'm okay with removing refine_amr but I want to get @jzuhone to sign off, too.

Sure ! I'm only removing because it single-handedly makes isolation impossible, as we discussed in private. Once the res

**Comment by jzuhone** [comment]
I'm not using it anymore really, so this is ok.

**Comment by neutrinoceros** [comment]
okay I think this one is now good for reviews !

**Comment by neutrinoceros** [comment]
rebased !

**Comment by brittonsmith** [comment]
Looks like a couple hangovers from my `halo_catalog` frontend PR. Let me know if you need any help sorting those out.

File: yt/frontends/halo_catalog/tests/test_outputs.py [hunk 1]
```diff
@@ -1,15 +1,9 @@
 import numpy as np
 
-from yt.convenience import load as yt_load
-from yt.frontends.halo_catalog.data_structures import YTHaloCatalogDataset
+from yt.frontends.halo_catalog.data_structures import HaloCatalogDataset
```
Comment (brittonsmith) [inline yt/frontends/halo_catalog/tests/test_outputs.py:3]:
```suggestion
from yt.frontends.halo_catalog.data_structures import YTHaloCatalogDataset
```

File: yt/frontends/halo_catalog/tests/test_outputs.py [hunk 2]
```diff
@@ -1,15 +1,9 @@
...
 from yt.frontends.ytdata.utilities import save_as_dataset
-from yt.testing import (
-    TempDirTest,
-    assert_array_equal,
-    assert_equal,
-    requires_file,
-    requires_module,
-)
+from yt.loaders import load as yt_load
+from yt.testing import TempDirTest, assert_array_equal, requires_module
```
Comment (brittonsmith) [inline yt/frontends/halo_catalog/tests/test_outputs.py:6]:
```suggestion
from yt.testing import (
    TempDirTest,
    assert_array_equal,
    assert_equal,
    requires_file,
    requires_module,
)
```

Comment (brittonsmith) [reply yt/frontends/halo_catalog/tests/test_outputs.py:6]: oops, this removes the `yt.loaders` line, too

Comment (brittonsmith) [inline yt/frontends/halo_catalog/tests/test_outputs.py:6]:
```suggestion
from yt.testing import (
    TempDirTest,
    assert_array_equal,
    assert_equal,
    requires_file,
    requires_module,
)
```

**Comment by neutrinoceros** [comment]
Thanks a lot for your assistance @brittonsmith , I appreciate it :)

**Comment by matthewturk** [comment]
More merge issues but this has two approvals, so once it's fixed up I think we can merge.

**Comment by neutrinoceros** [comment]
I'm on it

**Comment by neutrinoceros** [comment]
should be good !

**Pull request merged**